File: src/server/backend/runtime/plugins/qwen3guard.py
"""Shared Qwen3Guard API adapter for server-side LLM safety plugins."""
from __future__ import annotations

import json
import re
import urllib.request
from typing import Any

from shared.schemas.context import RuntimeContext
from shared.schemas.decisions import GuardDecision
from shared.schemas.events import RuntimeEvent

from backend.runtime.plugins.base import BasePlugin, CheckResult

DEFAULT_MODEL = "Qwen3Guard-Gen-8B"


class Qwen3GuardPluginBase(BasePlugin):
    policy_scope = "llm"

    def _messages_for_event(self, event: RuntimeEvent) -> list[dict[str, str]]:
        raise NotImplementedError

    def check(
        self,
        event: RuntimeEvent,
        context: RuntimeContext,
        trajectory_window: list[RuntimeEvent] | None = None,
    ) -> CheckResult:
        messages = self._messages_for_event(event)
        if not messages:
            return CheckResult.empty()

        api_url = str(getattr(self, "api_url", "") or "").strip()
        api_key = str(getattr(self, "api_key", "") or "").strip()
        if not api_url or not api_key:
            return CheckResult(
                is_final=False,
                metadata={"qwen3guard": {"error": "missing_api_config"}},
            )

        try:
            content = self._post_chat_completion(messages)
        except Exception as exc:
            return CheckResult(
                is_final=False,
                metadata={"qwen3guard": {"error": str(exc)}},
            )

        parsed = _parse_qwen3guard_content(content)
        metadata = {
            "qwen3guard": {
                "safety": parsed["safety"],
                "categories": parsed["categories"],
                "raw_content": content,
                "scope": self.policy_scope,
            }
        }
        safety = parsed["safety"]
        if safety == "safe":
            return CheckResult(is_final=False, metadata=metadata)
        if safety == "unsafe":
            signal = "qwen3guard_unsafe"
            return CheckResult(
                decision_candidate=GuardDecision.deny(
                    f"Qwen3Guard classified {self.policy_scope} content as unsafe.",
                    policy_id=f"server:qwen3guard:{self.policy_scope}:unsafe",
                    risk_signals=[signal],
                    metadata=metadata,
                ),
                risk_signals=[signal],
                is_final=True,
                metadata=metadata,
            )

        signal = "qwen3guard_controversial" if safety == "controversial" else "qwen3guard_unknown"
        return CheckResult(
            decision_candidate=GuardDecision.human_check(
                f"Qwen3Guard classified {self.policy_scope} content as {safety}; human review required.",
                policy_id=f"server:qwen3guard:{self.policy_scope}:{safety}",
                risk_signals=[signal],
                metadata=metadata,
            ),
            risk_signals=[signal],
            is_final=True,
            metadata=metadata,
        )

    def _post_chat_completion(self, messages: list[dict[str, str]]) -> str:
        api_url = str(getattr(self, "api_url", "") or "").strip()
        api_key = str(getattr(self, "api_key", "") or "").strip()
        model = str(getattr(self, "model", "") or DEFAULT_MODEL).strip()
        timeout_s = float(getattr(self, "timeout_s", 20.0) or 20.0)
        body = json.dumps({"model": model, "messages": messages}).encode("utf-8")
        request = urllib.request.Request(
            api_url,
            data=body,
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {api_key}",
            },
            method="POST",
        )
        with urllib.request.urlopen(request, timeout=timeout_s) as response:
            payload = json.loads(response.read().decode("utf-8"))
        return str(payload["choices"][0]["message"]["content"])


def _parse_qwen3guard_content(content: str) -> dict[str, Any]:
    safety_match = re.search(r"(?im)^Safety\s*:\s*([A-Za-z_-]+)", content or "")
    safety = safety_match.group(1).strip().lower() if safety_match else "unknown"
    categories_match = re.search(r"(?im)^Categories\s*:\s*(.+)$", content or "")
    raw_categories = categories_match.group(1).strip() if categories_match else ""
    categories = [
        item.strip()
        for item in re.split(r"[,;]", raw_categories)
        if item.strip() and item.strip().lower() != "none"
    ]
    if safety not in {"safe", "unsafe", "controversial"}:
        safety = "unknown"
    return {"safety": safety, "categories": categories}

File: src/client/js/agentguard/plugins/common/qwen3guard.js
"use strict";

const { CheckResult } = require("../base");
const { GuardDecision } = require("../../schemas/decisions");

const DEFAULT_MODEL = "Qwen3Guard-Gen-8B";
const DEFAULT_TEMPERATURE = 0.0;

function textOf(value) {
  if (value == null) {
    return "";
  }
  if (typeof value === "string") {
    return value;
  }
  if (typeof value === "number" || typeof value === "boolean" || typeof value === "bigint") {
    return String(value);
  }
  if (Array.isArray(value)) {
    return value.map(textOf).filter(Boolean).join("\n");
  }
  if (typeof value !== "object") {
    return String(value);
  }
  if (typeof value.text === "string") {
    return value.text;
  }
  if (typeof value.content === "string") {
    return value.content;
  }
  if (typeof value.output === "string") {
    return value.output;
  }
  if (typeof value.message === "string") {
    return value.message;
  }
  if (Array.isArray(value.content)) {
    return textOf(value.content);
  }
  try {
    return JSON.stringify(value);
  } catch (_) {
    return String(value);
  }
}

function parseQwen3GuardContent(content) {
  const source = String(content || "");
  const safetyMatch = source.match(/^Safety\s*:\s*([A-Za-z_-]+)/im);
  const categoriesMatch = source.match(/^Categories\s*:\s*(.+)$/im);
  let safety = safetyMatch ? String(safetyMatch[1] || "").trim().toLowerCase() : "unknown";
  if (!["safe", "unsafe", "controversial"].includes(safety)) {
    safety = "unknown";
  }
  const categories = categoriesMatch
    ? String(categoriesMatch[1] || "")
      .split(/[,;]/)
      .map((item) => item.trim())
      .filter((item) => item && item.toLowerCase() !== "none")
    : [];
  return { safety, categories };
}

async function postChatCompletion(plugin, messages) {
  const apiUrl = String(plugin.api_url || plugin.apiUrl || "").trim();
  const apiKey = String(plugin.api_key || plugin.apiKey || "").trim();
  const model = String(plugin.model || DEFAULT_MODEL).trim();
  const temperature = Number.isFinite(Number(plugin.temperature))
    ? Number(plugin.temperature)
    : DEFAULT_TEMPERATURE;
  const timeoutS = Number.isFinite(Number(plugin.timeout_s || plugin.timeoutS))
    ? Number(plugin.timeout_s || plugin.timeoutS)
    : 20.0;
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), timeoutS * 1000);
  try {
    const response = await fetch(apiUrl, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${apiKey}`,
      },
      body: JSON.stringify({ model, messages, temperature }),
      signal: controller.signal,
    });
    if (!response.ok) {
      throw new Error(`HTTP ${response.status}`);
    }
    const payload = await response.json();
    return String(payload?.choices?.[0]?.message?.content || "");
  } finally {
    clearTimeout(timeout);
  }
}

function resultFromContent(content, policyScope) {
  const parsed = parseQwen3GuardContent(content);
  const metadata = {
    qwen3guard: {
      safety: parsed.safety,
      categories: parsed.categories,
      raw_content: content,
      scope: policyScope,
    },
  };
  if (parsed.safety === "safe") {
    return new CheckResult({ is_final: false, metadata });
  }
  if (parsed.safety === "unsafe") {
    const signal = "qwen3guard_unsafe";
    return new CheckResult({
      decision_candidate: GuardDecision.deny(
        `Qwen3Guard classified ${policyScope} content as unsafe.`,
        {
          policy_id: `local:qwen3guard:${policyScope}:unsafe`,
          risk_signals: [signal],
          metadata,
        },
      ),
      risk_signals: [signal],
      is_final: true,
      metadata,
    });
  }
  const signal = parsed.safety === "controversial" ? "qwen3guard_controversial" : "qwen3guard_unknown";
  return new CheckResult({
    decision_candidate: GuardDecision.human_check(
      `Qwen3Guard classified ${policyScope} content as ${parsed.safety}; human review required.`,
      {
        policy_id: `local:qwen3guard:${policyScope}:${parsed.safety}`,
        risk_signals: [signal],
        metadata,
      },
    ),
    risk_signals: [signal],
    is_final: true,
    metadata,
  });
}

async function checkWithQwen3Guard(plugin, messages, policyScope) {
  if (!messages.length) {
    return CheckResult.empty();
  }
  const apiUrl = String(plugin.api_url || plugin.apiUrl || "").trim();
  const apiKey = String(plugin.api_key || plugin.apiKey || "").trim();
  if (!apiUrl || !apiKey) {
    return new CheckResult({
      is_final: false,
      metadata: { qwen3guard: { error: "missing_api_config" } },
    });
  }
  try {
    const content = await postChatCompletion(plugin, messages);
    return resultFromContent(content, policyScope);
  } catch (error) {
    return new CheckResult({
      is_final: false,
      metadata: { qwen3guard: { error: String(error && error.message ? error.message : error) } },
    });
  }
}

module.exports = {
  checkWithQwen3Guard,
  parseQwen3GuardContent,
  textOf,
};
